### PR TITLE
Added Shroomloin Variations and Warpbeetle Chorus Variation.

### PR DESCRIPTION
**Pull request by 0rc1nus**
- [ ] Added shaking feature (just like when curing zombie villager)

### CODE REVIEW

File: src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java [hunk 1]
```diff
@@ -66,7 +74,49 @@ protected void registerGoals() {
...
+    }
+
+    @Override
+    public void readAdditional(CompoundNBT compound) {
+        super.readAdditional(compound);
+        this.setVariant(compound.getBoolean("variant"));
+    }
+
+
+    public boolean getVariant() {
```
Comment (SwanX1) [inline src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java]: Should be renamed to `isChorus`

File: src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java [hunk 2]
```diff
@@ -66,7 +74,49 @@ protected void registerGoals() {
...
+	    return this.dataManager.get(CHORUS);
+    }
+
+    public void setVariant(boolean variant) {
```
Comment (SwanX1) [inline src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java]: Should be renamed to `setChorus`

File: src/main/java/com/nekomaster1000/infernalexp/events/MiscEvents.java [hunk 3]
```diff
@@ -75,25 +78,63 @@ public static void onModConfigEvent(final ModConfig.ModConfigEvent event) {
...
+                if (state.equals(Blocks.RED_MUSHROOM_BLOCK.getDefaultState())) {
+                    if (((ShroomloinEntity) entity).getFungusType() == 4) {
+                        shroomloinEntity.becomeAngryAt(event.getPlayer());
+                    }
+                }
+                if (state.equals(Blocks.BROWN_MUSHROOM_BLOCK.getDefaultState())) {
+                    if (((ShroomloinEntity) entity).getFungusType() == 5) {
+                        shroomloinEntity.becomeAngryAt(event.getPlayer());
+                    }
                 }
```
Comment (SwanX1) [inline src/main/java/com/nekomaster1000/infernalexp/events/MiscEvents.java]: Checking for the fungus type first would be more performant.

**Review by SwanX1 — DISMISSED** [review]

**Comment by SwanX1** [comment]
I'm going to wait until the shaking feature is implemented before merging.

File: src/main/java/com/nekomaster1000/infernalexp/events/MiscEvents.java [hunk 4]
```diff
@@ -75,25 +78,65 @@ public static void onModConfigEvent(final ModConfig.ModConfigEvent event) {
...
+            if (entity instanceof ShroomloinEntity) {
+                ShroomloinEntity shroomloinEntity = (ShroomloinEntity) entity;
+
+                if (((ShroomloinEntity) entity).getFungusType() == 1) {
+                    if (state.equals(Blocks.CRIMSON_NYLIUM.getDefaultState())
+                        || state.equals(Blocks.CRIMSON_STEM.getDefaultState())
+                        || state.equals(Blocks.STRIPPED_CRIMSON_STEM.getDefaultState())
+                        || state.equals(Blocks.WEEPING_VINES.getDefaultState())
+                        || state.equals(Blocks.WEEPING_VINES_PLANT.getDefaultState())
+                        || state.equals(Blocks.NETHER_WART_BLOCK.getDefaultState())) {
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/events/MiscEvents.java]: Please turn all of these groups of blocks into tags and use state.getBlock().isIn(Tag)

**Review by Dariensg — CHANGES_REQUESTED** [review]

File: src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java [hunk 5]
```diff
@@ -34,6 +41,9 @@
 
 public class WarpbeetleEntity extends CreatureEntity {
     private int attackTimer;
+    private int converstionTicks;
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java]:
```suggestion
    private int conversionTicks;
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java [hunk 6]
```diff
@@ -26,16 +30,19 @@
...
+    private static final DataParameter<Integer> FUNGUS_TYPE = EntityDataManager.createKey(ShroomloinEntity.class, DataSerializers.VARINT);
 	private static final DataParameter<Integer> STATE = EntityDataManager.createKey(ShroomloinEntity.class, DataSerializers.VARINT);
 	private static final DataParameter<Boolean> IGNITED = EntityDataManager.createKey(ShroomloinEntity.class, DataSerializers.BOOLEAN);
+	private static final DataParameter<Boolean> CONVERTING = EntityDataManager.createKey(ShroomloinEntity.class, DataSerializers.BOOLEAN);
 	private int lastActiveTime;
 	private int timeSinceIgnited;
 	private final int fuseTime = 59;
-
-	// public static final Ingredient TEMPTATION_ITEMS =
+    private int converstionTicks;
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java]:
```suggestion
    private int conversionTicks;
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java [hunk 7]
```diff
@@ -51,13 +58,107 @@ protected void registerData() {
...
+
+    @Override
+    public void readAdditional(CompoundNBT compound) {
+        super.readAdditional(compound);
+        this.setFungusType(compound.getInt("fungusType"));
+        this.setConversionTime(compound.getInt("ShroomloinConversion"));
+    }
+
+    private void setConversionTime(int time) {
+        this.converstionTicks = time;
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java]:
```suggestion
        this.conversionTicks = time;
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java [hunk 8]
```diff
@@ -51,13 +58,107 @@ protected void registerData() {
...
+    public int getFungusType() {
+	    return this.dataManager.get(FUNGUS_TYPE);
+    }
+
+    @Override
+    protected ActionResultType getEntityInteractionResult(PlayerEntity playerIn, Hand hand) {
+        ItemStack stack = playerIn.getHeldItem(hand);
+            if (this.getFungusType() != 1 && stack.getItem() == Items.CRIMSON_FUNGUS) {
+                this.setFungusType(1);
+                this.converstionTicks = 40;
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java]:
```suggestion
                this.conversionTicks = 40;
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java [hunk 9]
```diff
@@ -51,13 +58,107 @@ protected void registerData() {
...
+                this.setConverting(true);
+                if (!playerIn.isCreative()) {
+                    stack.shrink(1);
+                }
+                return ActionResultType.SUCCESS;
+            } else if (this.getFungusType() != 2 && stack.getItem() == Items.WARPED_FUNGUS) {
+                this.setFungusType(2);
+                this.converstionTicks = 40;
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java]:
```suggestion
                this.conversionTicks = 40;
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java [hunk 10]
```diff
@@ -51,13 +58,107 @@ protected void registerData() {
...
+                this.setConverting(true);
+                if (!playerIn.isCreative()) {
+                    stack.shrink(1);
+                }
+                return ActionResultType.SUCCESS;
+            } else if (this.getFungusType() != 3 && stack.getItem() == IEBlocks.LUMINOUS_FUNGUS.get().asItem()) {
+                this.setFungusType(3);
+                this.converstionTicks = 40;
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java]:
```suggestion
                this.conversionTicks = 40;
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java [hunk 11]
```diff
@@ -51,13 +58,107 @@ protected void registerData() {
...
+                this.setConverting(true);
+                if (!playerIn.isCreative()) {
+                    stack.shrink(1);
+                }
+                return ActionResultType.SUCCESS;
+            } else if (this.getFungusType() != 4 && stack.getItem() == Items.RED_MUSHROOM) {
+                this.setFungusType(4);
+                this.converstionTicks = 40;
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java]:
```suggestion
                this.conversionTicks = 40;
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java [hunk 12]
```diff
@@ -51,13 +58,107 @@ protected void registerData() {
...
+                this.setConverting(true);
+                if (!playerIn.isCreative()) {
+                    stack.shrink(1);
+                }
+                return ActionResultType.SUCCESS;
+            } else if (this.getFungusType() != 5 && stack.getItem() == Items.BROWN_MUSHROOM) {
+                this.setFungusType(5);
+                this.converstionTicks = 40;
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java]:
```suggestion
                this.conversionTicks = 40;
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java [hunk 13]
```diff
@@ -51,13 +58,107 @@ protected void registerData() {
...
+        return super.getEntityInteractionResult(playerIn, hand);
+    }
+
+    /**
 	 * Called to update the entity's position/logic.
 	 */
 	public void tick() {
 		if (this.isAlive()) {
+		    if (this.isConverting() && this.converstionTicks > 0) {
+		        this.converstionTicks--;
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java]:
```suggestion
		        this.conversionTicks--;
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java [hunk 14]
```diff
@@ -51,13 +58,107 @@ protected void registerData() {
...
+		        if (this.converstionTicks == 0) {
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java]:
```suggestion
		        if (this.conversionTicks == 0) {
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java [hunk 15]
```diff
@@ -51,13 +58,107 @@ protected void registerData() {
...
+
+    public boolean isShaking() {
+	    return this.isConverting();
+    }
+
+    @Override
+    public void writeAdditional(CompoundNBT compound) {
+        super.writeAdditional(compound);
+        compound.putInt("fungusType", this.getFungusType());
+        compound.putInt("ShroomloinConversionTime", this.converstionTicks);
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/ShroomloinEntity.java]:
```suggestion
        compound.putInt("ShroomloinConversionTime", this.conversionTicks);
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java [hunk 16]
```diff
@@ -66,14 +76,95 @@ protected void registerGoals() {
...
+
+    public boolean isShaking() {
+        return this.isConverting();
+    }
+
+    @Override
+    public void writeAdditional(CompoundNBT compound) {
+        super.writeAdditional(compound);
+        compound.putBoolean("chorus", this.isChorus());
+        compound.putInt("WarpbeetleConversionTime", this.converstionTicks);
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java]:
```suggestion
        compound.putInt("WarpbeetleConversionTime", this.conversionTicks);
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java [hunk 17]
```diff
@@ -66,14 +76,95 @@ protected void registerGoals() {
...
+
+    @Override
+    public void readAdditional(CompoundNBT compound) {
+        super.readAdditional(compound);
+        this.setChorus(compound.getBoolean("chorus"));
+        this.setConversionTime(compound.getInt("WarpbeetleConversionTime"));
+    }
+
+    private void setConversionTime(int time) {
+        this.converstionTicks = time;
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java]:
```suggestion
        this.conversionTicks = time;
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java [hunk 18]
```diff
@@ -66,14 +76,95 @@ protected void registerGoals() {
...
+    public void setChorus(boolean chorus) {
+	    this.dataManager.set(CHORUS, chorus);
+    }
+
+    @Override
+    protected ActionResultType getEntityInteractionResult(PlayerEntity playerIn, Hand hand) {
+        ItemStack stack = playerIn.getHeldItem(hand);
+        if (!this.isChorus() && stack.getItem() == Items.CHORUS_FRUIT) {
+            this.setChorus(true);
+            this.converstionTicks = 40;
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java]:
```suggestion
            this.conversionTicks = 40;
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java [hunk 19]
```diff
@@ -66,14 +76,95 @@ protected void registerGoals() {
...
+            this.setConverting(true);
+            if (!playerIn.isCreative()) {
+                stack.shrink(1);
+            }
+            return ActionResultType.SUCCESS;
+        }
+        else if (this.isChorus() && stack.getItem() == Items.WARPED_FUNGUS) {
+            this.setChorus(false);
+            this.converstionTicks = 40;
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java]:
```suggestion
            this.conversionTicks = 40;
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java [hunk 20]
```diff
@@ -66,14 +76,95 @@ protected void registerGoals() {
...
+    @Override
 	public void livingTick() {
 		super.livingTick();
 
         if (this.attackTimer > 0) {
             --this.attackTimer;
         }
 
+        if (this.isAlive()) {
+            if (this.isConverting() && this.converstionTicks > 0) {
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java]:
```suggestion
            if (this.isConverting() && this.conversionTicks > 0) {
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java [hunk 21]
```diff
@@ -66,14 +76,95 @@ protected void registerGoals() {
...
+                this.converstionTicks--;
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java]:
```suggestion
                this.conversionTicks--;
```

File: src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java [hunk 22]
```diff
@@ -66,14 +76,95 @@ protected void registerGoals() {
...
+                if (this.converstionTicks == 0) {
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/entities/WarpbeetleEntity.java]:
```suggestion
                if (this.conversionTicks == 0) {
```

File: src/main/java/com/nekomaster1000/infernalexp/events/MiscEvents.java [hunk 23]
```diff
@@ -75,25 +79,50 @@ public static void onModConfigEvent(final ModConfig.ModConfigEvent event) {
...
+        BlockState state = event.getState();
+        List<?> list = event.getPlayer().world.getEntitiesWithinAABB(ShroomloinEntity.class,
+            event.getPlayer().getBoundingBox().grow(32.0D));
+        for (int j = 0; j < list.size(); j++) {
+            Entity entity = (Entity) list.get(j);
+            if (entity instanceof ShroomloinEntity) {
+                ShroomloinEntity shroomloinEntity = (ShroomloinEntity) entity;
+
+                if (((ShroomloinEntity) entity).getFungusType() == 1) {
+                    if (state.getBlock().isIn(IETags.Blocks.ANGER_CRIMSON_BLOCKS)) {
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/events/MiscEvents.java]: Could you please refactor these tags to be ANGER_CRIMSON_SHROOMLOIN_BLOCKS, etc.

File: src/main/java/com/nekomaster1000/infernalexp/init/IETags.java [hunk 24]
```diff
@@ -15,6 +15,11 @@
         public static final ITag.INamedTag<Block> BASE_STONE_SHORES = tag("base_stone_shores");
         public static final ITag.INamedTag<Block> GLOW_FIRE_BASE_BLOCKS = tag("glow_fire_base_blocks");
         public static final ITag.INamedTag<Block> MAGMA_CUBE_AVOID_BLOCKS = tag("magma_cube_avoid_blocks");
+        public static final ITag.INamedTag<Block> ANGER_CRIMSON_BLOCKS = tag("anger_crimson_blocks");
```
Comment (Hellion-Games) [inline src/main/java/com/nekomaster1000/infernalexp/init/IETags.java]: Same as above, please refactor these tags to include "shroomloin"

**Review by Dariensg — CHANGES_REQUESTED** [review]

**Review by Dariensg — APPROVED** [review]

**Pull request merged**